### ISSUE TITLE
Thread starvation

### ISSUE THREAD

**Issue by kixelated** [tokio-rs/tokio-uring#80]
I saw #78 and I triggered flashbacks to an issue I ran into when implementing my own io_uring runtime a year ago.

I had 1k frontend requests evenly split between threads and all blocked (parked) on 1 backend fetch. When the backend fetch finishes, the thread that issued it is woken and can process its frontend requests. However, the requests bound to the other threads will stay blocked on iouring until they get randomly woken by unrelated network activity. This can cause significant latency spikes depending on the traffic load.

Basically, once a thread calls `submit_and_wait`, there's no way to unpark it for non-IO work.

My solution was to was create an `eventfd` for each thread. Right before getting parked, the thread queues up a read on the eventfd and submits all of the tasks to uring. In order to unpark a thread, another thread writes to the eventfd using its local uring. There's some atomics sprinkled in to avoid unnecessary wakeups.

My repo is currently private at the moment because my Rust sucks but I can make it public if it would help.

**Comment by Noah-Kennedy** [comment]
We currently park with epoll, which wakes when uring has completions available. This means we don't actually have this issue, although it *is* something we need to keep in mind whenever we move tokio-uring into mainline tokio.

**Comment by kixelated** [comment]
Ah apologies, I jumped to conclusions when I saw the comment in the PR. I agree that this shouldn't be an issue then.

Just to clarify, you're using epoll with the uring's FD instead of `submit_and_wait`. It does mean you have to issue two syscalls to park a thread (submit and then epoll) but you still get the benefit of batching with uring.

My runtime polls futures until they're all pending and then calls `submit_and_wait` to submit/park the thread. I'd love to see this implemented in tokio so I can delete my awful code!

**Comment by Noah-Kennedy** [comment]
@kixelated you are correct about our use of epoll. Also, something along the lines of what you are doing is the plan once we move tokio-uring into tokio (may be a while though).

By the way, a pro tip: periodically call submit_and_wait with a timeout of 0 on a set interval so that you don't starve tasks on your runtime. This is especially true if you use futures that aren't dependent on io_uring for waking (for example, async channels), as those can starve tasks you have which are dependent on the submit call.

**Comment by kixelated** [comment]
Yeah, the plan was to call `submit` if the uring queue reached a certain size or some duration had elapsed. I also wanted to investigate `IORING_SETUP_SQPOLL` as an alternative.
